File: 78_Subsets.cpp
#include<bits/stdc++.h>
using namespace std;

class Solution {
public:
    vector<vector<int>> ans;
    int n;
    vector<vector<int>> subsets(vector<int>& nums) {
        n=nums.size();
        addSubsets(nums, 0);

        return ans;
    }

    void addSubsets(vector<int>& nums, int i, vector<int> subset = vector<int>()) {
        if (i==n){
            ans.push_back(subset);
            return;
        }

        addSubsets(nums, i+1,subset);
        subset.push_back(nums[i]);
        addSubsets(nums, i+1,subset);


    }
};

int main(){
    Solution s;
    
    vector<int> in = {1,2,3};
    s.subsets(in);
    
    for (vector<int> &i : s.ans){
        for (int &j: i){
            cout<<j<<" ";
        }
        cout<<"\n";
    }
    return 0;
}

File: 78_Subsets.py
from typing import List

# DFS
# tc=O(n*(2^n)), sc = O(n) [for ouput = 2^n subsets and n*(2^(n-1)) integers in output]
class Solution:
    def subsets(self, nums: List[int]) -> List[List[int]]:
        n = len(nums)
        ans = []

        stack = []
        def dfs(i):
            if i==n:
                ans.append(stack[:])
                return
                
            dfs(i+1)

            stack.append(nums[i])
            dfs(i+1)
            stack.pop()

        dfs(0)
        return ans


# Iterative
# tc=O(n*(2^n)), sc = O(1) [for ouput = 2^n subsets and n*(2^(n-1)) integers in output]
class Solution:
    def subsets(self, nums: List[int]) -> List[List[int]]:
        ans = [[]]

        for num in nums:
            ans += [ subset + [num] for subset in ans]

        return ans